server/src-test/Data/TimeSpec.hs

```diff
@@ -1,39 +1,63 @@
 module Data.TimeSpec (spec) where
 -- | Time-related properties we care about.
 
-import Prelude
-import Data.Time.Clock.Units
-import Data.Time
-import Data.Aeson
-import Test.Hspec
+import           Data.Aeson
+import           Data.Time
+import           Data.Time.Clock.Units
+import           Prelude
+import           Test.Hspec
 
 spec :: Spec
 spec = do
-  timeUnitsSpec
+  timeAbsoluteUnitsSpec
+  timeCalendarUnitsSpec
   diffTimeSpec
 
-timeUnitsSpec :: Spec
-timeUnitsSpec =
-  describe "time units" $ do
+timeAbsoluteUnitsSpec :: Spec
```

Comment:
I think you could make this spec generic to avoid duplicating the code, given the changes I’ve suggested to `Data.Time.Clock.Units`.

Reply:
I'm stuck here 😅 

I have tried doing 

```haskell
class TimeUnitsSpec t where
  timeUnitSpec :: t -> Spec
  timeUnitSpec t =
    describe "time absolute units" $ do
      it "converts correctly" $ do
        (seconds 123 ::  t) `shouldBe` 123
        (milliseconds 123 ::  t) `shouldBe` 0.123
        (microseconds 123 ::  t) `shouldBe` 0.000123
        (nanoseconds 123 ::  t) `shouldBe` 0.000000123

      it "has a correct Read instance" $ do
        (seconds (read "123") ::  t) `shouldBe` 123
        (milliseconds (read "123") ::  t) `shouldBe` 0.123
        (microseconds (read "123") ::  t) `shouldBe` 0.000123
        (nanoseconds (read "123") ::  t) `shouldBe` 0.000000123

deriving instance TimeUnitsSpec (Duration 'Absolute)
deriving instance TimeUnitsSpec (Duration 'Calendar)
```

I'm not sure if this is how I should be doing, any pointers would be really great

Reply:
I realized while writing this comment that this is trickier than I originally anticipated, so don’t feel bad for not figuring it out. The solution isn’t especially complicated if you understand all the pieces, but it requires applying several distinct type-level programming building blocks, many of which are well beyond beginner-level Haskell!

In any case, you don’t need a typeclass here. Rather, we can get away with a polymorphic function:

```haskell
{-# LANGUAGE AllowAmbiguousTypes #-}

spec :: Spec
spec = do
  describe "absolute" $ timeUnitSpec @'Absolute
  describe "calendar" $ timeUnitSpec @'Calendar

timeUnitSpec
  :: forall t
   . ( Eq (Duration t)
     , Fractional (Duration t)
     , Show (Duration t)
     , Read (Duration t)
     ) => Spec
timeUnitSpec = do
  it "converts correctly" $ do
    seconds @t 123 `shouldBe` 123
    milliseconds @t 123 `shouldBe` 0.123
    microseconds @t 123 `shouldBe` 0.000123
    nanoseconds @t 123 `shouldBe` 0.000000123

  it "has a correct Read instance" $ do
    seconds @t (read "123") `shouldBe` 123
    milliseconds @t (read "123") `shouldBe` 0.123
    microseconds @t (read "123") `shouldBe` 0.000123
    nanoseconds @t (read "123") `shouldBe` 0.000000123
```

What’s going on here? A few things:

  1. We define a polymorphic function `timeUnitSpec` that is parameterized by some type `t`.

  2. We use an explicit `forall` to bring the `t` type variable into scope in the body using `ScopedTypeVariables`.

  3. In the body, we use `TypeApplications` to instantiate functions like `seconds` and `milliseconds` at `t`. You can think of `t` as a type-level *argument* to `timeUnitSpec`, and writing `seconds @t` passes that argument along to `seconds`.

  4. We need a bag of typeclass constraints on the function because GHC can’t figure out on its own that they hold for all possible choices of `t`. (Usually GHC will tell you if you’re missing any of these, so you should be able to just follow the suggestions in the type error.)

  5. Finally, we need to enable `AllowAmbiguousTypes` to get GHC to accept `timeUnitSpec`. Since `t` only appears in the *constraints* of `timeUnitSpec`’s type, GHC would normally have no way of figuring out what it’s supposed to be when it gets called—it could pick anything!

     However, we avoid that problem by using `TypeApplications` again to pass `t` explicitly, writing `timeUnitSpec @'Absolute` and `timeUnitSpec @'Calendar`, so the ambiguity is okay. Therefore, we ask GHC to be more permissive and allow the ambiguous type.

Let me know if you have any questions about all of this! I think the individual pieces aren’t too complicated in isolation, but figuring out how to put them all together in the right way is very non-obvious if you haven’t been exposed to them.

Reply:
Thank you so much for the explanation, I was not familiar with many of the concepts used her like `polymorphic functions`,`ScopedTypeVairables`,`TypeApplications`. I will be reading up further on these topics to understand them better.